controllers/capabilities/backend_controller.go

```diff
@@ -174,3 +202,71 @@ func (r *BackendReconciler) validateSpec(backendResource *capabilitiesv1beta1.Ba
 		FieldErrorList: errors,
 	}
 }
+
+func (r *BackendReconciler) removeBackend(backend *capabilitiesv1beta1.Backend) error {
+	providerAccount, err := controllerhelper.LookupProviderAccount(r.Client(), backend.Namespace, backend.Spec.ProviderAccountRef, r.Logger())
+	if err != nil {
+		return err
+	}
+
+	threescaleAPIClient, err := controllerhelper.PortaClient(providerAccount)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve all product CRs that are under the same ns as the backend CR
+	opts := k8sclient.ListOptions{
+		Namespace: backend.Namespace,
+	}
+	productCRsList := &capabilitiesv1beta1.ProductList{}
+	err = r.Client().List(context.TODO(), productCRsList, &opts)
+	if err != nil {
+		return err
+	}
+
+	// fetch CRs that belong to a tenant
+	tenantProductCRs, err := r.fetchTenantProductCRs(productCRsList, backend)
+
+	// update backendUsages for each product retrieved
+	for _, productCR := range tenantProductCRs {
+		delete(productCR.Spec.BackendUsages, backend.Spec.SystemName)
```

Comment:
Product CR has more references to the backend than in the backendUsages. I propose a method of the Product struct, to do that. Something like:
```go
func (product *Product) RemoveBackendRefs(backendSystemName string) bool
```

That method will remove, if there exists, references from:
* Application plan pricing rules
* Application plan limits
* Backend Usages

Will return `true` when a change has been made, `fatse` otherwise.

Reply:
Hey, so it looks to me that the application plan limits and pricing mentions of the backend do not block the backend from deletion and get cleaned up by 3scale. Would you still have it deleted manually anyway?

Reply:
Yes, but the product reconciler will put the Product CR into inconsistent state because it has references to not existing backend resources